File: apps/web/src/shared/api/codexResets.ts
/**
 * Codex usage-limit resets, read straight from codex-resets.com.
 *
 * This used to go through the home API's `/api/codex-resets`, which proxied
 * the same provider and normalised its answer. The deployed web app has no API
 * behind `localhost:8000`, so the dashboard card never loaded in production.
 * The provider is public and answers any origin (`Access-Control-Allow-Origin:
 * *`), so the browser asks it directly and the normalisation lives here. Its
 * `ETag` revalidation is the browser's HTTP cache's job now, and the minute-long
 * reuse is React Query's (`useCodexResets`).
 *
 * Only executed resets enter the snapshot: `scheduled_reset` and `active_watch`
 * are forecasts, not events, and are ignored exactly as the API ignored them.
 */

export type CodexResetSource = {
  type: 'x_post' | 'observed';
  author?: string;
  url: string;
};

export type CodexReset = {
  id: string;
  resetType: 'regular' | 'banked';
  announcedAt: string;
  text: string;
  source: CodexResetSource;
};

export type CodexResetsResponse = {
  source: 'codex-resets.com';
  fetchedAt: string;
  generatedAt: string;
  stale: boolean;
  latestReset: CodexReset | null;
  stats: {
    total: number;
    avgIntervalDays: number | null;
    longestIntervalDays: number | null;
  };
  resets: CodexReset[];
};

const UPSTREAM = 'https://codex-resets.com/api/v1';
const PAGE_SIZE = 100;
const MAX_PAGES = 100;
const TIMEOUT_MS = 20_000;
const DAY_MS = 86_400_000;

/** The provider answered with something the card cannot trust. */
export class CodexResetsError extends Error {
  readonly code: 'rate-limited' | 'upstream-error' | 'unreachable' | 'invalid-payload';

  constructor(code: CodexResetsError['code'], message: string) {
    super(message);
    this.name = 'CodexResetsError';
    this.code = code;
  }
}

function invalid(message: string): never {
  throw new CodexResetsError('invalid-payload', message);
}

function object(value: unknown, field: string): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    invalid(`${field} must be an object`);
  }
  return value as Record<string, unknown>;
}

/** An ISO timestamp that names its zone, back as UTC without milliseconds. */
function timestamp(value: unknown, field: string): string {
  if (typeof value !== 'string') invalid(`${field} must be an ISO timestamp`);
  // A zoneless timestamp would be read in the viewer's own zone and move every
  // reset by their offset, so it is refused rather than guessed at.
  if (!/(Z|[+-]\d{2}:?\d{2})$/i.test(value)) invalid(`${field} needs a timezone`);
  const time = Date.parse(value);
  if (!Number.isFinite(time)) invalid(`${field} is not a valid timestamp`);
  return new Date(time).toISOString().replace(/\.\d{3}Z$/, 'Z');
}

function source(value: unknown): CodexResetSource {
  const item = object(value, 'source');
  if (item.type !== 'x_post' && item.type !== 'observed') invalid('source.type is invalid');
  if (typeof item.url !== 'string' || !item.url.startsWith('https://')) {
    invalid('source.url is invalid');
  }
  if (item.author !== undefined && item.author !== null && typeof item.author !== 'string') {
    invalid('source.author is invalid');
  }
  return typeof item.author === 'string'
    ? { type: item.type, author: item.author, url: item.url }
    : { type: item.type, url: item.url };
}

function codexReset(value: unknown): CodexReset {
  const item = object(value, 'reset');
  const resetType = item.reset_type ?? item.resetType;
  if (typeof item.id !== 'string' || !item.id) invalid('reset.id is invalid');
  if (resetType !== 'regular' && resetType !== 'banked') invalid('reset.reset_type is invalid');
  if (typeof item.text !== 'string') invalid('reset.text is invalid');
  return {
    id: item.id,
    resetType,
    announcedAt: timestamp(item.announced_at ?? item.announcedAt, 'reset.announced_at'),
    text: item.text,
    source: source(item.source),
  };
}

function optionalNumber(value: unknown, field: string): number | null {
  if (value === undefined || value === null) return null;
  if (typeof value !== 'number' || !Number.isFinite(value)) invalid(`stats.${field} is invalid`);
  return value;
}

/** The longest gap between two consecutive resets, in days. */
function longestInterval(resets: readonly CodexReset[]): number | null {
  if (resets.length < 2) return null;
  const times = resets.map((item) => Date.parse(item.announcedAt)).sort((a, b) => a - b);
  let longest = 0;
  for (let index = 1; index < times.length; index += 1) {
    longest = Math.max(longest, (times[index] - times[index - 1]) / DAY_MS);
  }
  return longest;
}

async function get(path: string, signal: AbortSignal): Promise<Record<string, unknown>> {
  let response: Response;
  try {
    response = await fetch(`${UPSTREAM}${path}`, {
      headers: { Accept: 'application/json' },
      signal,
    });
  } catch (error) {
    if (signal.aborted) throw error;
    throw new CodexResetsError('unreachable', 'Codex Resets is unreachable');
  }
  if (response.status === 429) {
    throw new CodexResetsError('rate-limited', 'Codex Resets rate-limited the request');
  }
  if (!response.ok) {
    throw new CodexResetsError('upstream-error', `Codex Resets returned HTTP ${response.status}`);
  }
  try {
    return object(await response.json(), 'response');
  } catch (error) {
    if (error instanceof CodexResetsError) throw error;
    return invalid('Codex Resets returned invalid JSON');
  }
}

/** The caller's cancellation, bounded by the same 20 s the API allowed. */
function bounded(signal?: AbortSignal): AbortSignal {
  const timeout = AbortSignal.timeout(TIMEOUT_MS);
  return signal ? AbortSignal.any([signal, timeout]) : timeout;
}

export async function fetchCodexResets(signal?: AbortSignal): Promise<CodexResetsResponse> {
  const within = bounded(signal);
  const status = await get('/status', within);
  const data = object(status.data, 'status.data');
  const meta = object(status.meta, 'status.meta');
  const stats = object(data.stats, 'status.stats');
  const latest = data.latest_reset;

  const byId = new Map<string, CodexReset>();
  const seenCursors = new Set<string>();
  let cursor: string | null = null;
  for (let page = 0; ; page += 1) {
    if (page === MAX_PAGES) invalid(`resets pagination exceeded ${MAX_PAGES} pages`);
    const query = new URLSearchParams({ limit: String(PAGE_SIZE), order: 'asc' });
    if (cursor !== null) query.set('cursor', cursor);
    const payload = await get(`/resets?${query}`, within);
    if (!Array.isArray(payload.data)) invalid('resets.data must be an array');
    for (const row of payload.data) {
      const item = codexReset(row);
      byId.set(item.id, item);
    }
    const pagination = object(payload.pagination, 'resets.pagination');
    if (pagination.has_more !== true) break;
    const next = pagination.next_cursor;
    if (typeof next !== 'string' || !next || seenCursors.has(next)) {
      invalid('resets pagination cursor is invalid');
    }
    seenCursors.add(next);
    cursor = next;
  }

  const resets = [...byId.values()].sort(
    (a, b) => Date.parse(a.announcedAt) - Date.parse(b.announcedAt) || a.id.localeCompare(b.id),
  );
  const total = stats.total;
  if (typeof total !== 'number' || !Number.isInteger(total) || total < 0) {
    invalid('stats.total is invalid');
  }

  return {
    source: 'codex-resets.com',
    fetchedAt: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
    generatedAt: timestamp(meta.generated_at, 'status.meta.generated_at'),
    stale: false,
    latestReset: latest === undefined || latest === null ? null : codexReset(latest),
    stats: {
      total,
      avgIntervalDays: optionalNumber(stats.avg_interval_days, 'average'),
      longestIntervalDays: longestInterval(resets),
    },
    resets,
  };
}

File: services/api/app/adapters/codex_resets.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, replace
from datetime import UTC, datetime
from itertools import pairwise
from typing import Any, Literal, cast

import httpx

from app.config import UPSTREAM_TIMEOUT_SECONDS

CODEX_RESETS_BASE_URL = "https://codex-resets.com/api/v1"
ResetType = Literal["regular", "banked"]
SourceType = Literal["x_post", "observed"]


class CodexResetsProviderError(Exception):
    def __init__(
        self,
        code: str,
        message: str,
        *,
        transient: bool,
        retry_after_seconds: int | None = None,
    ) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.transient = transient
        self.retry_after_seconds = retry_after_seconds


class CodexResetsPayloadError(CodexResetsProviderError):
    def __init__(self, message: str) -> None:
        super().__init__("invalid-payload", message, transient=False)


def _iso(value: datetime) -> str:
    return value.astimezone(UTC).isoformat().replace("+00:00", "Z")


def _datetime(value: object, field: str) -> datetime:
    if not isinstance(value, str):
        raise CodexResetsPayloadError(f"{field} must be an ISO timestamp")
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as exc:
        raise CodexResetsPayloadError(f"{field} is not a valid timestamp") from exc
    if parsed.tzinfo is None:
        raise CodexResetsPayloadError(f"{field} needs a timezone")
    return parsed.astimezone(UTC)


def _object(value: object, field: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise CodexResetsPayloadError(f"{field} must be an object")
    return cast(dict[str, Any], value)


@dataclass(frozen=True, slots=True)
class CodexResetSource:
    type: SourceType
    url: str
    author: str | None = None

    def to_wire(self) -> dict[str, str]:
        result = {"type": self.type, "url": self.url}
        if self.author is not None:
            result["author"] = self.author
        return result

    @classmethod
    def from_wire(cls, value: object) -> CodexResetSource:
        source = _object(value, "source")
        source_type = source.get("type")
        if source_type not in {"x_post", "observed"}:
            raise CodexResetsPayloadError("source.type is invalid")
        url = source.get("url")
        if not isinstance(url, str) or not url.startswith("https://"):
            raise CodexResetsPayloadError("source.url is invalid")
        author = source.get("author")
        if author is not None and not isinstance(author, str):
            raise CodexResetsPayloadError("source.author is invalid")
        return cls(type=cast(SourceType, source_type), url=url, author=author)


@dataclass(frozen=True, slots=True)
class CodexReset:
    id: str
    reset_type: ResetType
    announced_at: datetime
    text: str
    source: CodexResetSource

    def to_wire(self) -> dict[str, object]:
        return {
            "id": self.id,
            "resetType": self.reset_type,
            "announcedAt": _iso(self.announced_at),
            "text": self.text,
            "source": self.source.to_wire(),
        }

    @classmethod
    def from_wire(cls, value: object) -> CodexReset:
        item = _object(value, "reset")
        reset_id = item.get("id")
        text = item.get("text")
        reset_type = item.get("reset_type", item.get("resetType"))
        announced_at = item.get("announced_at", item.get("announcedAt"))
        if not isinstance(reset_id, str) or not reset_id:
            raise CodexResetsPayloadError("reset.id is invalid")
        if reset_type not in {"regular", "banked"}:
            raise CodexResetsPayloadError("reset.reset_type is invalid")
        if not isinstance(text, str):
            raise CodexResetsPayloadError("reset.text is invalid")
        return cls(
            id=reset_id,
            reset_type=cast(ResetType, reset_type),
            announced_at=_datetime(announced_at, "reset.announced_at"),
            text=text,
            source=CodexResetSource.from_wire(item.get("source")),
        )


@dataclass(frozen=True, slots=True)
class CodexResetStats:
    total: int
    avg_interval_days: float | None
    longest_interval_days: float | None

    def to_wire(self) -> dict[str, int | float | None]:
        return {
            "total": self.total,
            "avgIntervalDays": self.avg_interval_days,
            "longestIntervalDays": self.longest_interval_days,
        }

    @classmethod
    def from_wire(cls, value: object) -> CodexResetStats:
        stats = _object(value, "stats")
        total = stats.get("total")
        average = stats.get("avgIntervalDays", stats.get("avg_interval_days"))
        longest = stats.get("longestIntervalDays")
        if isinstance(total, bool) or not isinstance(total, int) or total < 0:
            raise CodexResetsPayloadError("stats.total is invalid")
        for field, number in (("average", average), ("longest", longest)):
            if number is not None and (
                isinstance(number, bool) or not isinstance(number, int | float)
            ):
                raise CodexResetsPayloadError(f"stats.{field} is invalid")
        return cls(
            total=total,
            avg_interval_days=None if average is None else float(average),
            longest_interval_days=None if longest is None else float(longest),
        )


@dataclass(frozen=True, slots=True)
class CodexResetsSnapshot:
    fetched_at: datetime
    generated_at: datetime
    etag: str | None
    stale: bool
    latest_reset: CodexReset | None
    stats: CodexResetStats
    resets: tuple[CodexReset, ...]

    def to_wire(self) -> dict[str, object]:
        return {
            "source": "codex-resets.com",
            "fetchedAt": _iso(self.fetched_at),
            "generatedAt": _iso(self.generated_at),
            "stale": self.stale,
            "etag": self.etag,
            "latestReset": None if self.latest_reset is None else self.latest_reset.to_wire(),
            "stats": self.stats.to_wire(),
            "resets": [item.to_wire() for item in self.resets],
        }

    @classmethod
    def from_wire(cls, value: object) -> CodexResetsSnapshot:
        root = _object(value, "snapshot")
        resets = root.get("resets")
        if not isinstance(resets, list):
            raise CodexResetsPayloadError("snapshot.resets must be an array")
        latest = root.get("latestReset")
        etag = root.get("etag")
        if etag is not None and not isinstance(etag, str):
            raise CodexResetsPayloadError("snapshot.etag is invalid")
        return cls(
            fetched_at=_datetime(root.get("fetchedAt"), "snapshot.fetchedAt"),
            generated_at=_datetime(root.get("generatedAt"), "snapshot.generatedAt"),
            etag=etag,
            stale=bool(root.get("stale", False)),
            latest_reset=None if latest is None else CodexReset.from_wire(latest),
            stats=CodexResetStats.from_wire(root.get("stats")),
            resets=tuple(CodexReset.from_wire(item) for item in resets),
        )


def _retry_after(response: httpx.Response) -> int | None:
    try:
        return max(0, int(response.headers.get("Retry-After", "")))
    except ValueError:
        return None


def _raise_for_provider(response: httpx.Response) -> None:
    if response.status_code == 429:
        raise CodexResetsProviderError(
            "rate-limited",
            "Codex Resets rate-limited the request",
            transient=True,
            retry_after_seconds=_retry_after(response),
        )
    if response.status_code >= 500:
        raise CodexResetsProviderError(
            "upstream-error", "Codex Resets is temporarily unavailable", transient=True
        )
    if not response.is_success:
        raise CodexResetsProviderError(
            "upstream-error",
            f"Codex Resets returned HTTP {response.status_code}",
            transient=False,
        )


async def _get(client: httpx.AsyncClient, path: str, **kwargs: object) -> httpx.Response:
    try:
        return await client.get(
            f"{CODEX_RESETS_BASE_URL}{path}",
            headers=cast(dict[str, str] | None, kwargs.get("headers")),
            params=cast(dict[str, str | int] | None, kwargs.get("params")),
            timeout=UPSTREAM_TIMEOUT_SECONDS,
        )
    except (httpx.TimeoutException, httpx.RequestError) as exc:
        raise CodexResetsProviderError(
            "unreachable", "Codex Resets is unreachable", transient=True
        ) from exc


def _json(response: httpx.Response) -> dict[str, Any]:
    try:
        return _object(response.json(), "response")
    except ValueError as exc:
        raise CodexResetsPayloadError("Codex Resets returned invalid JSON") from exc


def _longest_interval(resets: tuple[CodexReset, ...]) -> float | None:
    if len(resets) < 2:
        return None
    ordered = sorted(item.announced_at for item in resets)
    return max((later - earlier).total_seconds() / 86_400 for earlier, later in pairwise(ordered))


async def fetch_codex_resets(
    client: httpx.AsyncClient,
    *,
    previous: CodexResetsSnapshot | None = None,
    now: Callable[[], datetime] = lambda: datetime.now(UTC),
) -> CodexResetsSnapshot:
    """Read executed resets only; scheduled/watch records never enter the snapshot."""
    headers = {"Accept": "application/json"}
    if previous is not None and previous.etag:
        headers["If-None-Match"] = previous.etag
    status_response = await _get(client, "/status", headers=headers)
    if status_response.status_code == 304:
        if previous is None:
            raise CodexResetsPayloadError("status returned 304 without cached data")
        return replace(previous, fetched_at=now().astimezone(UTC), stale=False)
    _raise_for_provider(status_response)

    status_payload = _json(status_response)
    status_data = _object(status_payload.get("data"), "status.data")
    meta = _object(status_payload.get("meta"), "status.meta")
    latest_value = status_data.get("latest_reset")
    latest = None if latest_value is None else CodexReset.from_wire(latest_value)
    stats_value = _object(status_data.get("stats"), "status.stats")

    by_id: dict[str, CodexReset] = {}
    cursor: str | None = None
    seen_cursors: set[str] = set()
    for _ in range(100):
        params: dict[str, str | int] = {"limit": 100, "order": "asc"}
        if cursor is not None:
            params["cursor"] = cursor
        response = await _get(
            client, "/resets", headers={"Accept": "application/json"}, params=params
        )
        _raise_for_provider(response)
        payload = _json(response)
        rows = payload.get("data")
        if not isinstance(rows, list):
            raise CodexResetsPayloadError("resets.data must be an array")
        for row in rows:
            item = CodexReset.from_wire(row)
            by_id[item.id] = item
        pagination = _object(payload.get("pagination"), "resets.pagination")
        if pagination.get("has_more") is not True:
            break
        next_cursor = pagination.get("next_cursor")
        if not isinstance(next_cursor, str) or not next_cursor or next_cursor in seen_cursors:
            raise CodexResetsPayloadError("resets pagination cursor is invalid")
        seen_cursors.add(next_cursor)
        cursor = next_cursor
    else:
        raise CodexResetsPayloadError("resets pagination exceeded 100 pages")

    resets = tuple(sorted(by_id.values(), key=lambda item: (item.announced_at, item.id)))
    total = stats_value.get("total")
    average = stats_value.get("avg_interval_days")
    stats = CodexResetStats.from_wire(
        {
            "total": total,
            "avg_interval_days": average,
            "longestIntervalDays": _longest_interval(resets),
        }
    )
    return CodexResetsSnapshot(
        fetched_at=now().astimezone(UTC),
        generated_at=_datetime(meta.get("generated_at"), "status.meta.generated_at"),
        etag=status_response.headers.get("ETag"),
        stale=False,
        latest_reset=latest,
        stats=stats,
        resets=resets,
    )
